include/deal.II/lac/trilinos_tpetra_vector.h

```diff
@@ -659,45 +677,75 @@ namespace LinearAlgebra
        */
       /** @{ */
       /**
-       * Compress the underlying representation of the Trilinos object, i.e.
-       * flush the buffers of the vector object if it has any. This function is
-       * necessary after writing into a vector element-by-element and before
-       * anything else can be done on it.
+       * Communicate all changes made to ghost entries of the vector
+       * (i.e., entries included in the locally relevant index set but not in
+       * the locally owned index set). The VectorOperation operation specifies
+       * how to deal with entries that multiple processors modified. This is
+       * necessary whenever we are done modifying the vector data on multiple
+       * (MPI) processes via add() or set().
+       *
+       * This function compresses the data structures and allows the vector to
+       * be used in all other operations, like matrix-vector products. This is
+       * a collective operation, i.e., it needs to be run on all processors
+       * when used in %parallel.
+       *
+       * After compress() was called, each processor can only modify locally
+       * owned elements. To modify entries belonging to a different processor
+       * again, one must first call decompress().
        *
        * See
        * @ref GlossCompress "Compressing distributed objects"
        * for more information.
        */
       void
-      compress(const VectorOperation::values operation);
+      compress(const VectorOperation::values operation,
+               const IndexSet               &locally_owned = IndexSet());
+
+      /**
+       * Prepare the vector data so that each processor can access the elements
+       * owned by itself and the ghost elements (relevant index set).
+       *
+       * This function scatters the data between the processors and allows the
+       * vector to be modified via add() and set(). This is a collective
+       * operation, i.e., it needs to be run on all processors when used in
+       * %parallel.
+       *
+       * After decompress() is called, each processor can modify locally owned
+       * and locally relevant elements. For the vector to be used in all other
+       * operations again, like matrix-vector products, one must first call
+       * compress().
+       */
+      void
+      decompress(const IndexSet &locally_relevant = IndexSet());
```

Comment:
I don't understand the need for this function. Isn't this what `compress()` already does?

The mental model I would like to stick with is the same one we use in the Epetra and PETSc wrappers: A vector is either (i) fully distributed, so does not store any ghost elements, or (ii) has ghost elements and is then read-only. In the latter case, the only way to create such a vector is by assignment (using `operator=`), so the `decompress()` operation you describe here would be part of `operator=`. If you wanted to break it into a function of its own, it should be a `private` one.

Reply:
> I don't understand the need for this function. Isn't this what compress() already does?

No, after compress() is called, there is no way to access ghost elements, not even for read-only purposes.
And internally compress() uses Tpetra::export() and decompress uses Tpetra::import()
And from the Tpetra Documentation:
"Where [Import](https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1Import.html) is for redistributing data from a uniquely-owned distribution to a possibly multiply-owned distribution. [Export](https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1Export.html) is for redistributing data from a possibly multiply-owned distribution to a uniquely-owned distribution."

Therefore, we have to be careful whether we call Tpetra::import() or if we call Tpetra::export().
This is also one of the problems with the original implementation of the copy-assignment operator. It always calls Tpetra::import(), even though it is possible one has to call Tpetra::export() instead. 
  
  
  
> The mental model I would like to stick with is the same one we use in the Epetra and PETSc wrappers: A vector is either (i) fully distributed, so does not store any ghost elements, or (ii) has ghost elements and is then read-only. In the latter case, the only way to create such a vector is by assignment (using operator=), so the decompress() operation you describe here would be part of operator=. If you wanted to break it into a function of its own, it should be a private one.

First of all, this mental model explains the original behavior of the copy-assignment operator. I take an other look into this and try to add this functionality into the copy-assignment operator. 
And I guess it is useful to keep the decompress() function, but I will turn it into a private one.

Reply:
Internal is fine.

Here is another part of the mental model we have been using: Since ghosted vectors are read-only, you can't write anything into it that would be communicated via `compress()`. In other words, it can't happen that you call `compress()`, find that all of the contributions for non-locally-owned (ghost) elements are gone, and then need to `decompress()` to get them back. For vectors that have ghost elements, you should never have to call `compress()`.

Reply:
See also #16378.

Reply:
After looking into this again, I came to the conclusion it is best to remove decompress(). So, the implementation is as close as possible to other vector classes in deal.II.